datastore/datastore.go

```diff
@@ -0,0 +1,322 @@
+		Metrics:         make([]MetricData, 0),
+		Events:          make([]EventData, 0),
+	}
+
+	c, projectId, err := NewClient()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	//add query metrics
+	kinds, err := DatastoreStatKindQueryResult(c.Dsc)
+	if err != nil {
+		log.Fatal(err)
+	}
+	result := DatastoreData(kinds, projectId)
+
+	for _, metricResult := range result {
+		data.Metrics = append(data.Metrics, metricResult)
+	}
+
+	//add stackdriver metrics
+	for _, metric := range stackdriverEndpoints {
+		resp, err := StackdriverResp(projectId, metric)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		result, err := StackdriverData(resp)
+		if err != nil {
+			log.Fatal(err)
+		}
+		for _, metricResult := range result {
+			data.Metrics = append(data.Metrics, metricResult)
+		}
+	}
+
+	fatalIfErr(log, helpers.OutputJSON(data, prettyPrint))
+}
+
+//NewClient() creates a client for datastore, it primarily exists for testing purposes
+func NewClient() (Client, string, error) {
+	dsClient, projectId, err := ConnectDatastore(base64Config)
+	if err != nil {
+		return Client{}, "", err
+	}
+
+	return Client{
+		dsClient,
+	}, projectId, nil
+}
+
+//DatastoreStatKindQueryResult performs a Query against the datastore and return values using __Stat_Kind__
+func DatastoreStatKindQueryResult(ds DatastoreClient) ([]DatastoreKind, error) {
+	q := datastore.NewQuery("__Stat_Kind__").Order("kind_name")
+
+	var kinds []DatastoreKind
+
+	ctx := context.Background()
+	_, err := ds.GetAll(ctx, q, &kinds)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return kinds, nil
+}
+
+//DatastoreData converts a []DatastoreKind to a []map[string]interface{} to be used for the final output to new relic
+func DatastoreData(kinds []DatastoreKind, projectID string) []map[string]interface{} {
+	var datastoreData []map[string]interface{}
+	for _, k := range kinds {
+		datastoreData = append(datastoreData, map[string]interface{}{
+			"event_type":                         "DatastoreSample",
+			"provider":                           "datastoreQuery",
+			"datastoreQuery.builtinIndexBytes":   k.BuiltinIndexBytes,
+			"datastoreQuery.builtinIndexCount":   k.BuiltinIndexCount,
+			"datastoreQuery.compositeIndexBytes": k.CompositeIndexBytes,
+			"datastoreQuery.compositeIndexCount": k.CompositeIndexCount,
+			"datastoreQuery.entityBytes":         k.EntityBytes,
+			"datastoreQuery.bytes":               k.Bytes,
+			"datastoreQuery.count":               k.Count,
+			"datastoreQuery.kindName":            k.KindName,
+			"datastoreQuery.projectId":           projectID,
+			"datastoreQuery.timestamp":           k.Timestamp.Unix(),
+		})
+	}
+	return datastoreData
+}
+
+func fatalIfErr(log *logrus.Logger, err error) {
+	if err != nil {
+		log.WithError(err).Fatal("can't continue")
+	}
+}
+
+// ConnectDatastore establishes a datastore.Client from a base64 encoding JSON credentials file.
+func ConnectDatastore(base64Config string) (*datastore.Client, string, error) {
```

Comment:
This looks to be a copy of the function from content-schema but with the projectId returned. Do you need the project ID? It is specific to the environment, ie dev, staging, production and this code only runs in one at a time. I think the environment is an easier identifier. Then again projectID is more exact, perhaps it would be best to modify the shared function to always output projectID as well to make it easier for this type of thing?

Reply:
I think that makes sense